### ISSUE TITLE
[BUG] Function App crashes on startup with error "Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData'"

### ISSUE THREAD

**Issue by JeroenJSmit** [Azure/azure-sdk-for-net#34467]
### Library name and version

Microsoft.Azure.WebJobs.Extensions.Storage 5.1.0

### Describe the bug

I am upgrading a Function App from using version 5.0.1 of Microsoft.Azure.WebJobs.Extensions.Storage to 5.1.0 and now the app crashes on startup with the following error:

Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

Looking at the release notes, which point to Microsoft.Azure.WebJobs.Extensions.Storage.Blobs, it seems to have to do with version 5.1.0 of that but even explicitely referencing that package the app crashes.

The exception is thrown in Microsoft.Azure.WebJobs.Host, DefaultExtensionRegistryFactory.cs registring the extension Microsoft.Azure.WebJobs.Extensions.Storage.Blobs.Config.BlobsExtensionConfigProvider

### Expected behavior

A minor version upgrade should not cause breaking changes nor a crash due to new dependencies

### Actual behavior

See description, it crashes.

### Reproduction Steps

I created a new Function App in VS 2022 (Version 17.4.5) and added a Blob trigger.
Upgraded all packages to the latest version, leading to the csproj as shown below.
It crashes on startup at the location mentioned in the description.

csproj content:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <AzureFunctionsVersion>v4</AzureFunctionsVersion>
    <UserSecretsId>89d3f15b-1b40-4c46-aedd-cf39b5aaf7f9</UserSecretsId>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Azure.Storage.Blobs" Version="12.15.0" />
    <PackageReference Include="Azure.Storage.Files.Shares" Version="12.13.0" />
    <PackageReference Include="Azure.Storage.Queues" Version="12.13.0" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="5.1.0" />
    <PackageReference Include="Microsoft.Extensions.Azure" Version="1.6.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="6.0.1" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="4.1.3" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

### Environment

**dotnet --info**

.NET SDK:
 Version:   7.0.103
 Commit:    276c71d299

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.103\

Host:
  Version:      7.0.3
  Architecture: x64
  Commit:       0a2bda10e8

.NET SDKs installed:
  2.0.3 [C:\Program Files\dotnet\sdk]
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.302 [C:\Program Files\dotnet\sdk]
  6.0.309 [C:\Program Files\dotnet\sdk]
  6.0.406 [C:\Program Files\dotnet\sdk]
  7.0.101 [C:\Program Files\dotnet\sdk]
  7.0.102 [C:\Program Files\dotnet\sdk]
  7.0.103 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.28 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Visual Studio: 17.4.5**

**Comment by gbpcor** [comment]
I'm running into this issue, too. Same upgrade.

**Comment by moehoward003** [comment]
Same issue here. I thought this was due to me trying to bind to Azurite.

**Comment by davide-bergamini-sevenit** [comment]
Same issue here

**Comment by amnguye** [comment]
Hi just acknowledging this is an issue and we're looking into this right now.

As an update I was able to reproduce all your issues with a simple app and got the same error.
![image](https://user-images.githubusercontent.com/48961492/220781969-c793d2d0-a0ab-4410-86b7-3a92a937b1cf.png)

**Comment by surensaluka** [comment]
Hi, downgrading to 5.0.1 seems to resolve the issue. (Clean the solution and run) Looking forward for a fix.

**Comment by angalbi** [comment]
Hi, I've downgraded Microsoft.Azure.WebJobs.Extensions.Storage 5.1.0 to  5.0.1, clean the project and it worked for me.

**Comment by diegorvieira17** [comment]
> Hi, I've downgraded Microsoft.Azure.WebJobs.Extensions.Storage 5.1.0 to 5.0.1, clean the project and it worked for me.

Hi, I've got the same problem described above. I tried to solve this by downgrading the package but the problem was not solved.

Any other advice?

Below are the package installed in my project:

![image](https://user-images.githubusercontent.com/7876106/220970803-2f13a2b1-4633-4a53-8ae2-820185d141f1.png)

**Comment by moehoward003** [comment]
> > Hi, I've downgraded Microsoft.Azure.WebJobs.Extensions.Storage 5.1.0 to 5.0.1, clean the project and it worked for me.
> 
> Hi, I've got the same problem described above. I tried to solve this by downgrading the package but the problem was not solved.
> 
> Any other advice?
> 
> Below are the package installed in my project:
> 
> ![image](https://user-images.githubusercontent.com/7876106/220970803-2f13a2b1-4633-4a53-8ae2-820185d141f1.png)

For verification, could you provide your error message? The original error was noted in VS 2022 (which is what I am using also) and seems you are using VS Code? Not sure if that matters, but if you are in fact getting the same error, could you try with VS 2022?

**Comment by diegorvieira17** [comment]
Thank's for your answer @moehoward003.

I'm using the Rider. Below are the VS 2022 screen error.

![image](https://user-images.githubusercontent.com/7876106/220984790-016a500c-01f3-4f15-b287-18bc44bf340f.png)

**Comment by moehoward003** [comment]
> Thank's for your answer @moehoward003.
> 
> I'm using the Rider. Below are the VS 2022 screen error.
> 
> ![image](https://user-images.githubusercontent.com/7876106/220984790-016a500c-01f3-4f15-b287-18bc44bf340f.png)
> 
> I searched for Microsoft.Azure.WebJobs, Version=3.0.34.0 in the package nuget page but I couldn't find it. Maybe this can be the problem.

Could you try removing the, "Microsoft.Azure.WebJobs.Extensions.Storage" package all together, compile to confirm you are missing a reference? (and based on that confirm 5.0.1 is installed?) And also maybe provide the code using the noted package? Not sure, just things to explore to perhaps uncover the issue. Let me know.

**Comment by amnguye** [comment]
We're looking to make a hotfix release to remove this error. It was a miscommunication on our part for releasing a new feature too early before another dependency has not rolled out their feature.

We apologize for any inconvenience this may have caused and I will update the thread once the hotfix release has been made.

**Comment by pragnagopa** [comment]
Apologies for the inconvenience. Please note new packages should not impact any existing applications in Azure. Dependencies have been rolled out. Azure Functions Core Tools - https://github.com/Azure/azure-functions-core-tools will be updated soon. 

Using latest Core Tools should allow users to upgrade to latest nuget packages.

**Comment by chrismimosil** [comment]
I have the same issue but have no reference to Microsoft.Azure.WebJobs.Extensions.Storage.  It happened when I upgraded Microsoft.Azure.WebJobs.Extensions.ServiceBus from 4.2.1 to latest (5.9.0) and that has no internal dependency on Microsoft.Azure.WebJobs.Extensions.Storage, but does depend on Microsoft.Azure.Webjobs (3.0.36).  Downgrading Microsoft.Azure.WebJobs.Extensions.ServiceBus to 5.8.1 fixes the issue

**Comment by liliankasem** [comment]
> I have the same issue but have no reference to Microsoft.Azure.WebJobs.Extensions.Storage. It happened when I upgraded Microsoft.Azure.WebJobs.Extensions.ServiceBus from 4.2.1 to latest (5.9.0) and that has no internal dependency on Microsoft.Azure.WebJobs.Extensions.Storage, but does depend on Microsoft.Azure.Webjobs (3.0.36). Downgrading Microsoft.Azure.WebJobs.Extensions.ServiceBus to 5.8.1 fixes the issue

It looks like the same thing happened in the ServiceBus extension where they also implemented changes to use `ParameterBindingData`. Anything consuming `ParameterBindingData` should be in preview. We're working to move these changes to a feature branch and release a hot fix to remove preview features from the stable packages.

In the meantime, using the latest Core Tools should unblock locally. There should be no impact in Azure.

**Comment by Tyler-V** [comment]
We encountered this as well, downgrading to 5.0.1 from 5.1.0 resolved the error

**Comment by Tyler-V** [comment]
@liliankasem this seems to be symptomatic with the issue that you cant specify what version of Core Tools you want to debug your Azure Functions with locally even if you have the latest installed

![image](https://user-images.githubusercontent.com/3893234/224828974-bed1f613-869e-4a90-9659-6f7076fcd9f3.png)

![image](https://user-images.githubusercontent.com/3893234/224828921-97593c40-b656-4c64-9986-a27687a93a83.png)

![image](https://user-images.githubusercontent.com/3893234/224829065-89738d86-a706-452c-85b6-e204813c057a.png)

If I remove the %AppData%/Local/AzureFunctionsTools folder and restart Visual Studio, it keeps on installing 4.32.0

**Comment by liliankasem** [comment]
I think this confirms that the error only shows up with versions prior to `4.0.5030` (which is expected with the current state of things, we can't change that). As for Visual Studio not using the latest core tools version, this isn't something I can help with - @pragnagopa do you know anything about that?

**Comment by DarinMacRae** [comment]
Chiming in here in case it helps.
@liliankasem wrote:
> There should be no impact to applications in Azure at all, this is just a local environment issue.

I'm not sure this is true. I deployed a function app to a staging environment yesterday that started logging these errors. I had just recently upgraded Microsoft.Azure.WebJobs.Extensions.Storage to 5.1.0.
I'm going to try taking this advice from @moehoward003 and see if I need to reference it at all.

> Could you try removing the, "Microsoft.Azure.WebJobs.Extensions.Storage" package all together,

**Comment by liliankasem** [comment]
> Chiming in here in case it helps. @liliankasem wrote:
> 
> > There should be no impact to applications in Azure at all, this is just a local environment issue.
> 
> I'm not sure this is true. I deployed a function app to a staging environment yesterday that started logging these errors. I had just recently upgraded Microsoft.Azure.WebJobs.Extensions.Storage to 5.1.0. I'm going to try taking this advice from @moehoward003 and see if I need to reference it at all.
> 
> > Could you try removing the, "Microsoft.Azure.WebJobs.Extensions.Storage" package all together,

What's the host version in the staging environment? Is it pinned at all? The host version that rolled out in January supports these changes

**Comment by DarinMacRae** [comment]
Currently, it shows:
```
"version": "4.14.0.19631",
"versionDetails": "4.14.0+fc14864b4fc2a095f8cd92c7f0ca338dbeabd430",
"platformVersion": "99.0.10.793"
```
I did not capture this information while it was logging the error though.
I have since, deployed the update that doesn't reference 5.1.0.

However, the exception detail did capture this which does imply version 4.14.
![image](https://user-images.githubusercontent.com/3474753/225972531-baa1804f-ce6c-418f-b6e1-13b674544b87.png)

**Comment by liliankasem** [comment]
The host version with support is 4.15.0+; I'm not sure how your host is still on 4.14.0 unless pinned? 🤔. The latest host that is out right now for windows is 4.16.1 and linux 4.16.2.

**Comment by DarinMacRae** [comment]
Both the staging slot (and the production slot, fwiw) are configured as such:
![image](https://user-images.githubusercontent.com/3474753/225976660-59fb7e9b-4ded-43fa-9c9c-483aefa9d1f0.png)

**Comment by DarinMacRae** [comment]
I have 6 instances of this function app running in two regions. (They are identical except for app settings.) Of those, only 2 are running 4.16. The other 4 are still on 4.14.

**Comment by aaronclausen** [comment]
I'm blocked by this.

I've installed the very latest Core Tools (4.0.5030). Closed and restarted VS, but when I debug my function locally, it's still using 4.0.4829 and so I'm still getting the "Could not load type...." issue related to Microsoft.Azure.WebJobs V 3.0.34.0

How can I remove 4.0.4829 and make VS use the latest Core Tools version?

**Comment by aaronclausen** [comment]
> I have the same issue but have no reference to Microsoft.Azure.WebJobs.Extensions.Storage. It happened when I upgraded Microsoft.Azure.WebJobs.Extensions.ServiceBus from 4.2.1 to latest (5.9.0) and that has no internal dependency on Microsoft.Azure.WebJobs.Extensions.Storage, but does depend on Microsoft.Azure.Webjobs (3.0.36). Downgrading Microsoft.Azure.WebJobs.Extensions.ServiceBus to 5.8.1 fixes the issue

Thanks a lot. Yep worked for me. Needed to clean and rebuild.

**Comment by eberthold** [comment]
> As for Visual Studio not using the latest core tools version, this isn't something I can help with

For Visual Studio you can update the core tools with this button in the options dialog which also resolved the problem for a colleague of mine. Because the installation of the standalone version isn't respected by Visual Studio

![image](https://user-images.githubusercontent.com/4233281/229816498-cdcac3c4-b619-418f-95d5-6058a8cdd1db.png)

**Comment by lachlanedmiston** [comment]
For anyone who is experiencing this only in their local environment: I partially resolved this issue by downgrading Microsoft.NET.Sdk.Functions to 4.1.2 and then upgrading back to 4.1.3.

When I debug locally in Visual Studio 2022 it's still running an old version of Core Tools (4.0.4829) despite newer being installed but it _is_ debugging and that's really all I need to continue.

For me the problem was caused by:
- manually installing func cli from .msi alongside whatever VS had done or
- upgrading a functions project from .NET Core 3.1 to .NET 6 and updating packages

I'd tried the solutions above with no luck.

**Comment by VMM-TG** [comment]
Hi,

just ran into this problem. My workaround:
- first I uninstalled all core tools (installed from msi and npm) to make sure only the one under AppData was available
- then I shutdown my Visual Studio 2022
- renamed AppData\Local\AzureFunctionsTools to AppData\Local\_AzureFunctionTools (just in case I needed to restore it)
- opened Visual Studio again
- Used the check update in VS as mentioned in [eberthold's comment](https://github.com/Azure/azure-sdk-for-net/issues/34467#issuecomment-1496026290) and then installed the updates
- Voilá, the correct version (.5095) was installed!

**Comment by amnguye** [comment]
I will be closing the issue for now since from the WebJobs Storage Extension SDK side of things, there are no actionable steps from the Storage Extension SDK side. The issue is more on the Azure Core Tools needing to be updated.

Feel free to keep commenting to continue discussing as versions get pinned down, so the versions match between the extension and the Azure Core tools.

**Comment by ravindrank** [comment]
> Hi,
> 
> just ran into this problem. My workaround:
> 
> * first I uninstalled all core tools (installed from msi and npm) to make sure only the one under AppData was available
> * then I shutdown my Visual Studio 2022
> * renamed AppData\Local\AzureFunctionsTools to AppData\Local_AzureFunctionTools (just in case I needed to restore it)
> * opened Visual Studio again
> * Used the check update in VS as mentioned in [eberthold's comment](https://github.com/Azure/azure-sdk-for-net/issues/34467#issuecomment-1496026290) and then installed the updates
> * Voilá, the correct version (.5095) was installed!

Thank you.. Following these steps resolved the issue for me. :)

**Comment by DomenicoFasano** [comment]
Hi, I suppose I have the same issue:
<img width="418" alt="image" src="https://user-images.githubusercontent.com/54447259/235911087-ae942e4f-e264-45ef-9d80-2c800f9771f7.png">
The difference is that I'm using Visual Studio Code and trying to launch python functions. I have no idea on what to do, any help will be appreciated :)

**Comment by nuckolls** [comment]
> 

Thanks! Your outlined processed is what worked for me. Nothing else did. Getting the core tools updated and getting VS to pick up the latest version was the key to getting past this error. None of the other solutions were working for me.

**Comment by DataBach-maker** [comment]
Is there any update on this ? I tried the recommended solutions. Unfortunately none worked out for me. I tried to name specific versions of the extensions, yet the error still occurs:

```
{
  "version": "2.0",
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": true,
        "excludedTypes": "Request"
      }
    }
  },
  "extensionBundle": {
    "id": "Microsoft.Azure.Functions.ExtensionBundle",
    "version": "[3.15.0, 4.0.0)",
    "extensions": {
      "ServiceBus": {
        "version": "3.0.33"
      },
      "Storage": {
        "version": "5.0.1"
      }
    }
  }
}
```

Above my `host.json`. I use Python and am trying to eventhub trigger.

**Comment by javiermtl** [comment]
Hi Guys.     
I'm getting the same error. I'm using VS Code, I'm not able to debug my function.    
how can I fix it. 
A host error has occurred during startup operation '3b6af1eb-9feb-4a9d-9275-998666e8007d'.
[2023-05-24T13:36:24.679Z] Microsoft.Azure.WebJobs.Extensions.Storage.Blobs: Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
Value cannot be null. (Parameter 'provider')

**Comment by fvalencia12** [comment]
I just had the same problem and un-installed the msi version by going to windows add/remove programs and removing the Azure  Core tools and then installed it in Visual Studio like @eberthold explains above. 

Thanks guys for your help!

**Comment by codebrane** [comment]
This is an issue with Visual Studio for Mac 17.5.7 (build 6). The function app cannot be run locally:

"Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'."

**Comment by fvalencia12** [comment]
This is running in Azure.

Fabian Valencia
Information Technology
William Blair
The William Blair Building | 150 North Riverside Plaza, Chicago, Illinois 60606
Direct: +1-312-364-5132
***@***.******@***.***> | williamblair.com<http://www.williamblair.com/>
Facebook<https://www.facebook.com/WilliamBlairCo/> | LinkedIn<https://www.linkedin.com/company/166939> | YouTube<https://www.youtube.com/channel/UC7CPnCaTlZcLxOF7WFwY0uw> | Twitter<https://twitter.com/WilliamBlair>

**Comment by KevinPoorDeveloper** [comment]
What a mess - I have no clue what to do. I've tried all of the above.

**Comment by KevinPoorDeveloper** [comment]
> This has since been addressed and should not be a an issue if you're using the latest core tools version

This is such a mess, I've wasted an entire day on trying all the solutions listed here. I'm using the newest core tools.
How am I supposed to move forward?

This is the bastard:
<PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.10.0" />

And it seems to only fail when I try and do cosmosdb stuff like this:

public void Run([CosmosDBTrigger(databaseName: "MerchandiseTransaction", containerName: "Users",             
            Connection = "CosmosMerchandiseConn",
            CreateLeaseContainerIfNotExists = true, LeaseContainerName ="leases")] IReadOnlyList<Document> input, FunctionContext context)

**Comment by fabiocav** [comment]
@KevinPoorDeveloper  can you share the Core Tools output showing the version information? Thanks!

**Comment by KevinPoorDeveloper** [comment]
@fabiocav -like dotnetinfo? I'm not sure how to do what you're asking. Thank you!

**Comment by KevinPoorDeveloper** [comment]
> Core Tools output showing the version information

Like this?

Azure Functions Core Tools
Core Tools Version:       4.0.4829 Commit hash: N/A  (64-bit)
Function Runtime Version: 4.11.2.19273

[2023-06-21T15:52:40.510Z] Found C:\Software\path\path\project.csproj. Using for user secrets file configuration.
[2023-06-21T15:52:41.099Z] A host error has occurred during startup operation '11473e68-dbcc-4e0e-b51a-421eb56b9eaa'.
[2023-06-21T15:52:41.100Z] Microsoft.Azure.WebJobs.Extensions.CosmosDB: Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
Value cannot be null. (Parameter 'provider')
Press any key to continue....

**Comment by KevinPoorDeveloper** [comment]
I can see my version of Core Tools is slightly off, and that the version number of core tools that I have installed in ONE PLACE isn't "respected" from another. Another image from above said to use this checker, in Visual Studio...Visual Studio thinks everything is wonderful. 

![image](https://github.com/Azure/azure-sdk-for-net/assets/83042774/936bd789-ebf9-47bd-af3a-15241679baf0)

**Comment by KevinPoorDeveloper** [comment]
Unless I'm doing this wrong I don't even know how this is possible - it's not in "add remove programs"

```
C:\Software\azurefunctions_project>npm uninstall -g azure-functions-core-tools

removed 35 packages in 517ms

C:\Software\azurefunctions_project>npm -v azure-fuctions-core-tools
9.5.1
```

**Comment by KevinPoorDeveloper** [comment]
Some awful combination of steps from the above solved my issues. It sucks that we're seeking help with this **HERE**. I guarantee if I open up a support ticket they'd have no clue how to fix this.

**Comment by fabiocav** [comment]
Yeah, apologies for the issue. Can you detail how you were testing/debugging? That will help us understand what tooling failed to update your environment. If this was an F5 debug failure from VS, that indicates VS had not downloaded the latest Core Tools version, otherwise (e.g., you were running `func`), it would be helpful to understand how you've installed Core Tools on your machine.

Thanks for the details, this helps!

**Comment by fvalencia12** [comment]
One of the issues in my opinion is that there is multiple ways to install it, Visual Studio, VS Code or MSI. If a combination of mechanisms is used, then even if one is upgraded and things get out of sync, environments become brittle and stop working. The team should try and standardize in the installs or find a mechanism that would allow all installs to detect what a healthy environment should be and clean other installs.

Regards

Fabian Valencia
Information Technology
William Blair
The William Blair Building | 150 North Riverside Plaza, Chicago, Illinois 60606
Direct: +1-312-364-5132
***@***.******@***.***> | williamblair.com<http://www.williamblair.com/>
Facebook<https://www.facebook.com/WilliamBlairCo/> | LinkedIn<https://www.linkedin.com/company/166939> | YouTube<https://www.youtube.com/channel/UC7CPnCaTlZcLxOF7WFwY0uw> | Twitter<https://twitter.com/WilliamBlair>

**Comment by fabiocav** [comment]
Thank you @fvalencia12 , this is something the team is actively tracking, but the feedback is very appreciated.

@amamounelsayed for awareness.

**Comment by KevinPoorDeveloper** [comment]
It's unclear to me how things got into the AppData folder, and if that is even the "correct" place. I ended up (I think) deleting every azure* nuget, removing core tools anyway i could, then install them from the VS options > environment "Check for Updates" button. 

I think various nugets might be installing the wrong version of core tools? I have no clue. At this point that's all I had done.

It's maddening that it looks like the problem was encountered, and then the best fix was to implement a button that wasn't capable of detecting the problem in VS. This is like a full workday spent on this.

**Comment by qd-danh** [comment]
Same problems as above for me. Azure function ran successfully when launched from command line (outside of VS) using `func host start` but failed with the same error `Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData` when launch from VS (either F5 debug, or without debugging).

Wasted almost a day trying lots of the fixes above, and ultimately came down to the version of Visual Studio vs the version of Azure Functions Core Tools.

Had installed latest Azure Functions Core Tools (v4.0.5198) and see that version confirmed in the command line approach - startup sequence. 

While I was getting the error when launching from VS, I used sysinternals process monitor to see that VS was launching `dotnet` pointed at the `func.dll` at `%localappdata%\AzureFunctionsTools\Releases\4.40.0\cli_x64\func.dll`  (OK, weird).  AND, that DLL has version that's BACK A FEW MINOR points `v4.0.3971.0` which matches the output I see when that func host starts up in the command line as launched by VS (then followed by errors).

![image](https://github.com/Azure/azure-sdk-for-net/assets/85181651/1cf49a0d-8785-4475-b18e-6f2babf96469)

So -- how / where / why is VS using this version of `func.dll` that is causing pain and suffering? 

I also noticed compared to others in this thread, my VS doesn't have `Azure Functions` node in Tools | Options...

![image](https://github.com/Azure/azure-sdk-for-net/assets/85181651/3e5bc5d4-7b7a-4afb-b091-721972d70eff)

Ultimately, THE FIX: Updated Visual Studio! 
- From: VS 2022 17.2.6
- To:   VS 2022 17.6.4

Now I can successfully run the function from VS, both F5 debug and without debugging. Not a single change to the code or references.

Hopefully all this info helps somebody else that's in pain.

/cc @fabiocav

**Comment by KevinPoorDeveloper** [comment]
I also have PTSD launching a function locally now.

**Comment by codebrane** [comment]
In case it's of use, I had this issue with a ServiceBusTrigger function. I updated the core tools but it did nothing to solve the issue. I could not find the referenced version of WebJobs anywhere. I had to downgrade Microsoft.Azure.WebJobs.Extensions.ServiceBus to 5.8.1 to get it to run locally. I then re-upgraded Microsoft.Azure.WebJobs.Extensions.ServiceBus to 5.11.0 again and the issue did not reappear.

**Comment by davek-eql** [comment]
> This is an issue with Visual Studio for Mac 17.5.7 (build 6). The function app cannot be run locally:
> 
> "Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'."

I am also experiencing it with VS for Mac, 17.6.3 (build 421).

System.Private.CoreLib: Exception has been thrown by the target of an invocation. Could not load type 'Microsoft.Azure.WebJobs.ParameterBindingData' from assembly 'Microsoft.Azure.WebJobs, Version=3.0.34.0, Culture=neutral

**Comment by thinkOfaNumber** [comment]
I'm posting this for anyone else following with the same unique set of circumstances as mine! (Since this ticket is now a bit of a howto).

I'm not using Visual Studio nor doing C# development. I'm using VSCode to develop standard logic apps. Despite doing so every day for the last couple of months, something broke today and I started getting this error - core tools hasn't been updated since I first installed it by MSI (func-cli-4.0.4865-x64.msi) in August.

Anyway I found out that the MS Learn page for getting started with logic apps ([Create a Standard logic app workflow in single-tenant Azure Logic Apps using Visual Studio Code](https://learn.microsoft.com/en-us/azure/logic-apps/create-single-tenant-workflows-visual-studio-code)) is directing people to install a specific version of Core Tools 4.0.4865.

I found a later version ([4.0.5441](https://github.com/Azure/azure-functions-core-tools/releases/tag/4.0.5441)), installed it, and now everything is back to normal.
